### ISSUE TITLE
Refactor of enketo-webform.js, enketo-webform-edit.js, enketo-webform-view.js

### ISSUE THREAD

**Issue by eyelidlessness** [enketo/enketo#1030]
### Motivation

The enketo-express app entrypoints (enketo-webform.js, enketo-webform-edit.js, enketo-webform-view.js) have been a particular focal point of regressions over recent months[^1]. This has been largely because:

1. They've lacked test coverage—this is addressed currently in enketo/enketo-express#304 and two subsequent branches for edit/view.
2. There are minor but important differences between offline-capable, online-only, edit and view modes; these differences are sometimes hard to remember or follow.

In review of enketo/enketo-express#304, some changes were requested that would produce significant rework. While there was general agreement that those changes would be warranted, we also discussed an alternative where we take advantage of the new test coverage to refactor these app entrypoints, to reduce the level of risk overall.

### Goals

- Take advantage of the improved test coverage in enketo/enketo-express#304 to make the app entrypoints more maintainable and easier to reason about.
- The app entrypoints should not directly implement complex or significant logic. They should instead defer to smaller, more testable, units. As those smaller unites are introduced, they will be tested directly _in addition to_ the backfilled integration tests, which will be either removed or pared down and improved at the end of the refactor process.
- Make small, granular changes. Each step will be performed in a pair of commits, the first introducing the refactored behavior with more focused tests, and the second replacing existing behavior with the backfill tests still passing in CI.
- Eliminate most of the redundancy between offline/online/edit/view. Minor differences between modes should be easily identifiable and understood, as branches within a single flow.
- Because these steps will be granular, this refactor will be a gradual process over time.
- Unless otherwise noted, the changes discussed below will be restricted to:
  - The entrypoint modules themselves
  - Changes specifically affecting related modules
  - Sparingly, in affected modules where appropriate

### Plan

1. Finish backfilling tests of edit/view. (This is actually done, branched from and awaiting approval of enketo/enketo-express#304.)
1. Restore test coverage reporting. This is partially done in enketo/enketo-express#304 as well, but will be removed from watch mode testing for performance reasons. While coverage reporting in general isn't a great metric, showing very high coverage of these modules during the refactor will hopefully provide confidence that the changes are safe. We _may_ also add a coverage reporting GitHub check (e.g. [this Codecov check](https://docs.codecov.com/docs/github-checks)) to aid review during the refactor.
1. Make desired formatting changes, specifically enketo/enketo-express#328. This will be performed project-wide to minimize commit history noise. While this isn't _technically_ relevant to the refactor, it will improve my confidence in performing subsequent steps. Ideally this would have been performed _first_, but I suspect rebasing the test backfill work would cause unnecessarily difficult merge conflicts.
1. These other high-value formatting changes will be performed gradually as existing code is touched/moved:

    - Move toward `async`/`await`
    - Inline `export` statements

1. Wrap any top-level behavior in functions which can be called by tests. Initializing the app on load will not be performed in tests, only when running the app.
1. Parameterize functions which currently depend on top-level, global or implicit state. This will improve the ability to reason about data flow and state, and the dependencies between different pieces of functionality.
1. Eliminate mutation/side-effecting behavior (e.g. treat function parameters as immutable) where possible, and isolate it otherwise (e.g. separating DOM mutation from other data flow).
1. Document the public API of enketo-express, as discussed in enketo/enketo-express#300. This is mostly a formality, allowing us to move some functionality to other modules where they better fit without implicitly expanding the public API.
1. Move logic out of the entrypoint modules to appropriate dedicated modules, where appropriate modules already exist. E.g.

    - `_showErrorOrAuthenticate` -> gui.js
    - `_setAppCacheEventHandlers` -> application-cache.js
    - `_setFormCacheEventHandlers` -> form-cache.js
    - `_setEmergencyHandlers` -> store.js[^2]

1. Move other cross-cutting concerns around survey business logic/data flow to a new survey.js module. For now, this includes:

    - `_initOffline` (introduced in enketo/enketo-express#304, previously called top-level on import)
    - `_initOnline` (introduced in enketo/enketo-express#304, previously called top-level on import)
    - `_addBranding`
    - `_swapTheme`
    - `_prepareInstance`
    - `_init` (renamed and exported as initController or whatever makes the most sense)

    Over time, this module can accumulate other survey-specific logic which is currently spread out over many other modules.
1. Unify online/offline/edit/view initialization logic as discussed under Goals, branching only for specific differences in their behavior.
1. Remove the now-redundant backfilled integration tests, or pare them down and make some of the most valuable improvements discussed in enketo/enketo-express#304.

### Post-mortem

This is a really big set of changes. I think it's worth doing. I think the test backfills make it possible to do safely. But I acknowledge that any refactor carries some risk. Once this process is complete (or if at any point we interrupt the work in progress), it will be a good idea to assess:

1. What went well
1. What didn't go so well
1. Any action items that can come from that
1. Any guidelines we can establish for future work

[^1]: As well as controller-webform.js, but that will be addressed separately. The modules discussed here have a lot of similarity which is why they're being addressed together.
[^2]: 1 I'd prefer not to put _more_ DOM logic in store.js, but at least as a first measure this seems the most appropriate place for that function, as written.
